pkg/operator/controller/ingress/load_balancer_service.go

```diff
@@ -159,3 +163,21 @@ func (r *reconciler) finalizeLoadBalancerService(ci *operatorv1.IngressControlle
 	}
 	return nil
 }
+
+// ensureLoadBalancerCleanupFinalizer ensures the "service.kubernetes.io/load-balancer-cleanup"
+// finalizer does not exist for the load balancer service of ic. For additional background on
+// this finalizer, see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190423-service-lb-finalizer.md
+func (r *reconciler) ensureLoadBalancerCleanupFinalizer(ic *operatorv1.IngressController) error {
```

Comment:
If this logic were moved inline to [finalizeLoadBalancerService()](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/load_balancer_service.go#L151) you could avoid a new function and call to `r.currentLoadBalancerService()` (and I think it makes sense anyway because we're literally delaying finalization of the LB service until some other condition is true)

Reply:
@ironcladlou I tried that initially, but then `ensureIngressDeleted()` never proceeds to remove other `ingresscontroller` dependent resources until the cloud infra is cleaned-up. With this approach, the final `ingresscontroller` reconciliation only needs to remove the "ingresscontroller.operator.openshift.io/finalizer-ingresscontroller" finalizer. Which approach do you prefer?

Reply:
Hm, is this back to @Miciah's point about failing too soon? To avoid dealing with graphs directly (a path we played with once), the easy thing I've learned to do (from @deads2k) is to make as much progress as possible with parents, aggregate errors, and retry.

Right now we do

```go
finalizeLoadBalancerService() // if fail return early
deleteWildcardDNSRecord() // if fail return early
ensureRouterDeleted() // if fail return early — but finalizeLoadBalancerService will never succeed until this is called!
// finalize ingresscontroller
```

Would this make things better?

```go
finalizeLoadBalancerService() // append errors
deleteWildcardDNSRecord() // append errors
ensureRouterDeleted() // append errors
if errors > 0 return // and retry
// finalize ingresscontroller
```

Reply:
@ironcladlou ptal at the latest commit that follows your guidance.